osu.Game.Rulesets.Osu/Statistics/AccuracyHeatmap.cs

```diff
@@ -0,0 +1,226 @@
+                        }
+                    },
+                    pointGrid = new GridContainer
+                    {
+                        RelativeSizeAxes = Axes.Both
+                    }
+                }
+            };
+
+            Vector2 centre = new Vector2(points_per_dimension) / 2;
+            float innerRadius = centre.X * inner_portion;
+
+            Drawable[][] points = new Drawable[points_per_dimension][];
+
+            for (int r = 0; r < points_per_dimension; r++)
+            {
+                points[r] = new Drawable[points_per_dimension];
+
+                for (int c = 0; c < points_per_dimension; c++)
+                {
+                    HitPointType pointType = Vector2.Distance(new Vector2(c, r), centre) <= innerRadius
+                        ? HitPointType.Hit
+                        : HitPointType.Miss;
+
+                    var point = new HitPoint(pointType)
+                    {
+                        Colour = pointType == HitPointType.Hit ? new Color4(102, 255, 204, 255) : new Color4(255, 102, 102, 255)
+                    };
+
+                    points[r][c] = point;
+                }
+            }
+
+            pointGrid.Content = points;
+
+            if (score.HitEvents == null || score.HitEvents.Count == 0)
+                return;
+
+            // Todo: This should probably not be done like this.
+            float radius = OsuHitObject.OBJECT_RADIUS * (1.0f - 0.7f * (score.Beatmap.BaseDifficulty.CircleSize - 5) / 5) / 2;
+
+            foreach (var e in score.HitEvents.Where(e => e.HitObject is HitCircle))
+            {
+                if (e.LastHitObject == null || e.PositionOffset == null)
+                    continue;
+
+                AddPoint(((OsuHitObject)e.LastHitObject).StackedEndPosition, ((OsuHitObject)e.HitObject).StackedEndPosition, e.PositionOffset.Value, radius);
+            }
+        }
+
+        protected void AddPoint(Vector2 start, Vector2 end, Vector2 hitPoint, float radius)
+        {
+            if (pointGrid.Content.Length == 0)
+                return;
+
+            double angle1 = Math.Atan2(end.Y - hitPoint.Y, hitPoint.X - end.X); // Angle between the end point and the hit point.
+            double angle2 = Math.Atan2(end.Y - start.Y, start.X - end.X); // Angle between the end point and the start point.
+            double finalAngle = angle2 - angle1; // Angle between start, end, and hit points.
+            float normalisedDistance = Vector2.Distance(hitPoint, end) / radius;
+
+            // Convert the above into the local search space.
+            Vector2 localCentre = new Vector2(points_per_dimension) / 2;
+            float localRadius = localCentre.X * inner_portion * normalisedDistance; // The radius inside the inner portion which of the heatmap which the closest point lies.
+            double localAngle = finalAngle + Math.PI - MathUtils.DegreesToRadians(rotation); // The angle inside the heatmap on which the closest point lies.
```

Comment:
See how the comments I've added feel. I'm not really good at explaining this stuff :(

Reply:
Hmm, I feel that the previous version was easier to grasp upon understanding the pi trick.

The way I would look at it is that the angles calculated go `end -> start` and `end -> actual hit point`. If we consider the line going through start and end:

* If the actual hit point was on the *near* side, the `end -> actual hit point` vector will have the same direction as `end -> start`, therefore the angle will be near zero, therefore it will end up on the right side of the coordinate origin (relative to the positive part of the X axis) - but we want it to be exactly opposite that.
* Conversely, if the hit is on the *far* side, it will create a 180 degree angle, ending up on the left, but we actually want it on the right. This is what the 180deg rotation (or pi in radians) is for.

Might just be me, though - interested in @peppy's take on this.